[MRG] Add Detection Error Tradeoff (DET) curve classification metrics

dmohns opened the pull request:
#### Reference Issues/PRs

Continuation of the awesome work in PR #4980 and PR #8724 respectively.

- Resumes and closes #4980, closes #8724


#### What does this implement/fix? Explain your changes.

Scope of this PR (and the PRs it is based on) is to implement the functionality
to compute Detection Error Tradeoff (DET) curves from within scikit-learn.
DET-curves can used as a classification evaluation metric in scenarios where
an explicit comparison of different errors types is required.


#### Any other comments?

Credit goes to the authors @jkarnows @jucor of the aforementioned PRs likewise.
For now I have essentially fixed some basic inconsistencies which made test fail
and prevented the doc from being compiled.

DET curves are a niche classification evaluation metric.
Moving forward I would like to know whether they have a place in Scikit and if contribution like this is appreciated before putting additional effort into it.

If so, I am happy to continue to work on this PR as per ToDo list below.

#### ToDo list and next steps (without particular ordering)

- [x] Fix broken/non-existent links in doc
- [x] Fix failing tests
- [x] Add tests for DET curves
- [x] Expand Usage notes with more background information about DET curves. Explain how DET curves can be used and be interpreted.
- [x] (Optional) Add example plot
- [x] Add automated invariance tests


#### Open questions

Additionally I compiled a list of question about the status-quo so far

- Do we need references in the module description also?
I compared to similar modules like ROC and Precision-Recall curves but the use of references is somewhat inconsistent. I personally would prefer to not put references in multiple places and opted to put them into the documentation.
- I could not find any documentation whether or not to add contributors to the list of authors. I removed the references for now, but am happy to add them back if desired.
- Any further comments of the original authors @jkarnows and @jucor ?
- Any further suggestion for the ToDo list?

jkarnows committed:
Initial add DET curve to classification metrics

jucor committed:
Add DET to exports

Commit:
Fix DET-curve doctest errors

- Sample snippet in  model_evaluation documentation was outdated.

dmohns committed:
Clarify wording in DET-curve computation

- Align to the wording of ranking module to make it consistent.
- Add correct describtion of input and outputs.
- Update and fix non-existent links

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -217,6 +217,80 @@ def _binary_uninterpolated_average_precision(
+                                   sample_weight=None):
+    """Compute error rates for different probability thresholds.
+
+    Note: This implementation is restricted to the binary classification task.
```
Is it commonly applied to non-binary tasks?

dmohns replied:
Not that I am aware of, but I am anything but an expert on that domain. By their nature of comparing error types they are best suited for evaluation of binary detection tasks.

jnothman replied:
Maybe you can reword so this does not seem like a fault of the implementation

jnothman commented:
All code contributors will be noted in the changelog entry, which goes in doc/whats_new/v0.20.rst.

Could you add an example plotting a det curve and illustrating its usage, perhaps in contrast to roc?

jnothman commented on doc/modules/model_evaluation.rst:
```diff
@@ -1193,6 +1194,45 @@ if the predicted outputs have been binarized.
+`detection error tradeoff curve, or DET curve <https://en.wikipedia.org/wiki/Detection_error_tradeoff>`_.
+Quoting Wikipedia :
+
+  "A detection error tradeoff (DET) graph is a graphical plot of error rates for binary classification systems, plotting false reject rate vs. false accept rate. The x- and y-axes are scaled non-linearly by their standard normal deviates (or just by logarithmic transformation), yielding tradeoff curves that are more linear than ROC curves, and use most of the image area to highlight the differenc ...
```
please limit line lengths to 80 characters. Blockquotes can be indented.

Could you add discussion of how to use this for model selection / evaluation  / critique?

dmohns committed:
Beautify DET curve documentation source

- Limit line length to 80 characters.

dmohns committed:
Expand DET curve documentation

- Add an example plot to show difference between ROC and DET curves.
- Expand Usage Note section with background information and properties
of DET curves.

dmohns commented:
I added a small example to compare ROC and DET curves on the same classification task amongst further documentation and usage notes about DET curves. Happy to receive feedback for both.

dmohns committed:
Update DET-curve documentation

- Fix typos and some grammar improvements.
- Use named references to avoid potential conflicts with other sections.
- Remove unneeded references and improved existing ones by using e.g.
using versioned links.

dmohns committed:
Select relevant DET points using slice object

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -283,12 +283,22 @@ def detection_error_tradeoff_curve(y_true, y_score, pos_label=None,
-    first_ind = fps[::-1].searchsorted(fps[0])
-    sl = range(first_ind, last_ind)[::-1]
-    return fps[sl] / n_count, fns[sl] / p_count, thresholds[sl]
+    sl = slice(first_ind, last_ind)
```
You can set the step so that the slice is reversed

dmohns replied:
I initially went that route. Unfortunately that makes the computation of the slice borders very opaque. To deal with the fact that the "upper" slice border is exclusive we would have subtract `-2` in total with additional exception handling for the cases `1` and `0`.
I figured no one would ever be able follow what is happening and opted for the explicit approach.

jnothman reviewed (approved):
Otherwise LGTM

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.20.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

dmohns committed:
Remove some dubiety from DET curve doc-string

dmohns committed:
Add DET curve contributors

dmohns commented:
I have two more questions regarding this PR:
- When I build the documentation locally the auto generated example will not hyperlink  `detection_error_tradeoff_curve` in the source code. Am I missing something or is this behavior expected?
- Would we need a dedicated unit test for this metric or is the implicit test through `doctest` and auto generated example sufficient in this case?

jnothman commented:
Oh no! I let you get away without a unit test? Yes you should try to test its main properties and it's support for edge cases (ties, perfect scores, all same prediction).

As to docs, you can see the rendered docs through circle ci artists (see our developer docs for details)

dmohns commented:
1. Ok, I have to admit I was a little surprised to see this approved before even finishing my to-do list. Will try to add unit tests for the cases mentioned.

2. After a little hassle to access the CircleCI documentation I could confirm the link *does work* on the documentation built by CircleCI.

jnothman commented:
Sorry I've been a bit hasty!

dmohns committed:
Add tests for DET curves

dmohns commented:
I added some basic tests oriented at other tests in `test_ranking`. Any other cases that should be tested?

Side question: Now that I am touching `ranking.py` and `test_ranking.py` anyways can I also PEP8 them? (There are just a few line breaks missing.)

jnothman commented:
Re pep8: Please do not change unrelated things. It makes your contribution harder to review and may introduce merge conflicts to other pull requests. If it's really only missing line breaks, the second issue may not be severe so you could submit a quick separate PR.

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -693,6 +694,100 @@ def test_score_scale_invariance():
+
+def test_detection_error_tradeoff_curve_toydata():
+    # Check on a batch of small examples.
+    test_cases = [
```
Please use pytest.mark.parametrize instead

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -693,6 +694,100 @@ def test_score_scale_invariance():
+        [0, 1, 1], [0, 0.5]
+    )
+
+    # When the true_y values are all the same a detection error tradeoff cannot
```
y_true

dmohns committed:
Streamline DET test by using parametrization

dmohns commented:
> Re pep8: Please do not change unrelated things.
Ok, makes a lot of sense. Will open a separate PR for those.

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -693,6 +696,107 @@ def test_score_scale_invariance():
+    (0), (0.25), (0.5), (0.75), (1)
+])
+def test_detection_error_tradeoff_curve_constant_scores(y_score):
+    # Check computation with perfect and constant scores.
```
you're not testing with perfect scores, are you?

dmohns replied:
Good catch, will add.

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -693,6 +696,107 @@ def test_score_scale_invariance():
+    ([1, 0, 1], [0.5, 0.75, 1], [1, 1, 0], [0, 0.5, 0.5]),
+    ([1, 0, 1], [0.25, 0.5, 0.75], [1, 1, 0], [0, 0.5, 0.5]),
+])
+def test_detection_error_tradeoff_curve_toydata(
```
We would more conventionally style this as:
```py
def test_detection_error_tradeoff_curve_toydata(y_true, y_score,
                                                expected_fpr, expected_fnr):
```

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -693,6 +696,107 @@ def test_score_scale_invariance():
+    ([1, 0, 1], [0.25, 0.5, 0.5], [1], [0]),
+    ([1, 1, 0], [0.25, 0.5, 0.5], [1], [0]),
+])
+def test_detection_error_tradeoff_curve_tie_handling(
```
as above

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -693,6 +696,107 @@ def test_score_scale_invariance():
+    # be computed.
+    with np.errstate(all="raise"):
+        assert_raises(
+            FloatingPointError,
```
would we be better off with a more explicit error message?

dmohns replied:
I was considering it, but decided against it. 

Currently the function behaves like that: When you call detection_error_tradeoff_curve with all `y_true` the same it will return `NaN` in one of the return values and NumPy will throw a Warning. I find this behavior sort-of intuitive, as it suggests that the issue might be in the data and not in the way the function is called.

But I have no strong feeling towards either solution. I can change it, if this is standard behavior in other scikit-learn functions.

jnothman replied:
My concern is that cases like this do happen, at least when the calling of the metric is automated within some cross-validation or similar routine (and hence data sub-samples are taken). If it is going to be hard for users to debug, we should give them a reasonably informative error message... if it doesn't involve excessive additional computation to handle their edge case.

dmohns replied:
Good point. Given the computational overhead is indeed very small I will make this a little more explicit.

jnothman reviewed (commented):
Nice work

jnothman commented:
Should this now be marked [MRG]?

dmohns committed:
Increase verbosity of DET curve error handling

- Explicitly sanity check input before computing a DET curve.
- Add test for perfect scores.
- Adapt indentation style to match the test module.

qinhanmin2014 reviewed (commented):
Please try to include the metric in test_common.py like other metrics and remove redundant tests if possible.

dmohns commented:
Interesting, I didn't realize there was automated invariance checking happening. Unfortunately it doesn't yet cover curves https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_common.py#L53

For now, I will add DET curve to the comment.

qinhanmin2014 commented:
> For now, I will add DET curve to the comment.

Thanks :) That's exactly what I mean.

dmohns committed:
Add reference for DET curves in invariance test

jnothman commented:
I suppose to ensure input checking, putting some arbitrary aggregate of the
curves in common tests would be good, actually?

dmohns commented:
I had a quick look at the test_common: Some of the tests do make sense for curve metrics in general and DET curve in particular, e.g. order invariance, pos_label etc... Decent input checking could probably be achieved using existing test without even the need to mess with arbitrary aggregates.

Funny side fact: Now that we raise an error when only one class is present in y_true this line in invariance check `y_true = random_state.randint(0, 2, size=(20, )` is going to give us some headaches. 

Regardless, adding curve support to test_common might (or might not) involve some major changes to the test_common module itself. I am not really sure if this can be in scope of this PR. Thoughts?

jnothman commented:
that's why we fix the random state!

jnothman commented:
I would very much enjoy a PR which extended the invariance tests for
metrics to curves, classification_report and confusion_matrix. I find those
invariance tests very helpful. it would make sense to separate it from this
PR.

dmohns commented:
Agreed. The invariance tests seem like a desirable thing to have for curves. 

Another thought crossed my mind: Maybe we can change all asserts in test_common to assert_arrays. Assuming assert_arrays work for scalar, which I think they do. That way we could simply reuse the already existing tests.

dmohns commented:
What is the best way to proceed with this? 

I managed to put something together that will make test_common support non-scalar metrics. It will probably need some more beautification but it seem to work for now. If you want I can share.

Should we put the DET curve PR on hold and try to add curves to invariance test first or vice-versa?

jnothman commented:
it might take some time for this to get a second review in any case. Make a
separate PR with common tests changes, and we'll take a look.

dmohns commented:
Should I already resolve the merge conflicts or wait for a review?

Commit:
Add automated invariance checks for DET curves

Commit:
Resolve merge artifacts

Commit:
Make doctest happy

Commit:
Fix whitespaces for doctest

dmohns commented:
I rebased to master, resolved all the conflicts and moved the changed log to `doc/whats_new/v0.21.rst`. Is this correct?

Also, I added automated invariance test using the implementations from #10730 as suggested by @qinhanmin2014 

@qinhanmin2014 @jnothman any further comments, feedback, suggestions?

jnothman commented on examples/model_selection/plot_det.py:
```diff
@@ -0,0 +1,121 @@
+
+# prepare plots
+
+# first plot the ROC curve
```
plot -> prepare

jnothman commented on examples/model_selection/plot_det.py:
```diff
@@ -0,0 +1,121 @@
+ax_roc.xaxis.set_major_formatter(
+    FuncFormatter(lambda y, _: '{:.0%}'.format(y)))
+
+# second plot the DET curve
```
plot -> prepare

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -725,6 +726,111 @@ def test_score_scale_invariance():
+def test_detection_error_tradeoff_curve_constant_scores(y_score):
+    fpr, fnr, threshold = detection_error_tradeoff_curve(
+        y_true=[0, 1, 0, 1, 0, 1],
+        y_score=y_score * np.ones(6)
```
use np.full

jnothman reviewed (commented):
otherwise LGTM

jnothman commented:
sorry for the additional nitpicks.

Commit:
Revert unintended whitespace changes

Commit:
Revert unintended white space changes #2

Commit:
Fix typos and grammar

Commit:
Fix white space in doc

Commit:
Streamline test code

dmohns commented:
Thank you @jnothman for the feedback. I addressed your comments in additional commits.

cmarmo commented:
@dmohns, thanks for your patience, you already have one approval: are you still interested in working on this PR? If yes do you mind fixing conflicts, then maybe @agramfort or @NicholasHug could have a look? Thanks!

agramfort reviewed (commented):
diff looks clean but I would need to see an artifact of the rendered doc to better evaluate the user facing part.

dmohns committed:
Remove rebase artifacts

dmohns committed:
Fix PR link in doc

dmohns commented:
@cmarmo @agramfort Thanks for having a look at this PR. I have not used scikit-learn in quite a while and did not closely follow changes in the lib. Fixing the merge conflicts was pretty messy, please review carefully.

dmohns committed:
Fix test_ranking

dmohns committed:
Fix rebase errors

dmohns committed:
Fix import

cmarmo commented on doc/modules/model_evaluation.rst:
```diff
@@ -1437,6 +1438,90 @@ to the given limit.
+     Wikipedia, The Free Encyclopedia. September 4, 2017, 23:33 UTC.
+     Available at: https://en.wikipedia.org/w/index.php?title=Detection_error_tradeoff&oldid=798982054.
+     Accessed February 19, 2018.
+  .. [Martin1997] A. Martin, G. Doddington, T. Kamm, M. Ordowski, and M. Przybocki,
```
Martin 1997 is not cited in the text: to avoid the sphinx warning you can either cite it where appropriate either remove the square bracket content.

cmarmo commented on sklearn/metrics/_ranking.py:
```diff
@@ -218,6 +218,82 @@ def _binary_uninterpolated_average_precision(
+    """Compute error rates for different probability thresholds.
+    Note: This metrics is used for ranking evaluation of a binary
+    classification task.
+    Read more in the :ref:`User Guide <det_curve>`.
```
```suggestion
    Read more in the :ref:`User Guide <det_curve>`.
    
```

cmarmo commented on sklearn/metrics/_ranking.py:
```diff
@@ -218,6 +218,82 @@ def _binary_uninterpolated_average_precision(
+    pos_label : int, optional (default=None)
+        The label of the positive class
+    sample_weight : array-like of shape = [n_samples], optional
+        Sample weights.
```
```suggestion
        Sample weights.
        
```

cmarmo commented on sklearn/metrics/_ranking.py:
```diff
@@ -218,6 +218,82 @@ def _binary_uninterpolated_average_precision(
+        rate of predictions with score >= thresholds[i]. This is occasionally
+        referred to as false rejection or miss rate.
+    thresholds : array, shape = [n_thresholds]
+        Decreasing score values.
```
```suggestion
        Decreasing score values.
        
```

cmarmo commented on sklearn/metrics/_ranking.py:
```diff
@@ -218,6 +218,82 @@ def _binary_uninterpolated_average_precision(
+    See also
+    --------
+    roc_curve : Compute Receiver operating characteristic (ROC) curve
+    precision_recall_curve : Compute precision-recall curve
```
```suggestion
    precision_recall_curve : Compute precision-recall curve
    
```

cmarmo reviewed (commented):
Hope this will help in fixing sphinx warnings and documentation rendering.

dmohns committed:
Bring back newlines

- Swallowed by copy/paste

dmohns committed:
Remove uncited ref link

cmarmo commented on examples/model_selection/plot_det.py:
```diff
@@ -0,0 +1,121 @@
+    )
+
+# finally add legend
+ax_det = plt.subplot(1, 2, 2)
```
Apparently this line generates a `DeprecationWarning` from matplotlib (see the [rendered example](https://113142-843222-gh.circle-artifacts.com/0/doc/auto_examples/model_selection/plot_det.html#sphx-glr-auto-examples-model-selection-plot-det-py)): it would be nice if the warning could be avoided. Thanks!

dmohns replied:
Will see what I can do, currently matplot lib is giving me a hard time ...

cmarmo commented:
Thanks @dmohns for syncing, I think this can be moved to the [MRG] status... :)
@agramfort if you can find some time to review this one... Thanks a lot!

agramfort commented on examples/model_selection/plot_det.py:
```diff
@@ -0,0 +1,121 @@
+
+In this example we compare receiver operating characteristic (ROC) and
+detection error tradeoff (DET) curves to demonstrate how DET curves can help
+to asses the performance of different classification algorithms.
```
```suggestion
to assess the performance of different classification algorithms.
```

agramfort commented on examples/model_selection/plot_det.py:
```diff
@@ -0,0 +1,121 @@
+DET curves are commonly plotted in normal deviate scale.
+To achieve this we transform the errors rates as returned by the
+``detection_error_tradeoff_curve`` function and the axis scale using
+``scipy.stats.norm``.
```
please add a few words on what this example demonstrates? what is the take home message? Are you saying that DET makes it easier to highlight that RF is better than ROC?

dmohns replied:
This example merely tries to demonstrate to the unfamiliar user how an example of a DET curve can look like. I tried to not repeat myself too much, and the properties, pros and cons are already discussed in the usage guide (which uses the sample plot).
Would you be fine if I added an explicit link to the usage guide here too?

agramfort reviewed (commented):
I just looked at example https://113142-843222-gh.circle-artifacts.com/0/doc/auto_examples/model_selection/plot_det.html to get convinced. Maybe you can make the point clearer in the example?

agramfort commented:
let me try to clarify what I suggest. When you say "can help to assess
...". Please add a sentence
why it helps as evidenced by the plot below? I am more asking "why" users
should consider DET
and this example should be a proof that it does help. Maybe a couple more
sentences commenting
the plot can help looking at the right thing. I am sure it's obvious to you
but not to anyone.

thanks

>

dmohns committed:
Remove matplotlib deprecation warning

dmohns committed:
Bring back hidden reference

dmohns committed:
Add motivation to DET example

dmohns committed:
Fix lint

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -1437,6 +1438,92 @@ to the given limit.
+  A. Martin, G. Doddington, T. Kamm, M. Ordowski, and M. Przybocki,
+  `The DET Curve in Assessment of Detection Task Performance
+  <http://www.dtic.mil/docs/citations/ADA530509>`_,
+  NIST 1997.
```
Why not formatting this ref like the 2 above?

cmarmo replied:
Because this ref was not used in the text and sphinx complained about the uncited reference (see [comment](https://github.com/scikit-learn/scikit-learn/pull/10591#discussion_r456885084))

dmohns replied:
Exactly, sphinx was complaining. However, I think it is an even better approach to actually cite this article in docs as it gives a good explanation of how DET curves are helpful in comparing the error type tradeoff.

agramfort commented on examples/model_selection/plot_det.py:
```diff
@@ -0,0 +1,145 @@
+# prepare plots
+
+# first prepare the ROC curve
+ax_roc = plt.subplot(1, 2, 1, label='roc_curve')
```
Use plt.subplots would be more modern matplotlib but maybe a nitpick here

dmohns committed:
Add citation

dmohns committed:
Use modern matplotlib API

cmarmo commented:
@scikit-learn/core-devs a two years old PR, double approved! :) Someone available for merging?

agramfort commented:
thx @dmohns

lorentzenchr commented:
Thank you, @dmohns, for this contribution.

Would it make sense to have a `plot_detection_error_tradeoff_curve` like `plot_roc_curve`?

glemaitre commented:
@lorentzenchr You can comment there https://github.com/scikit-learn/scikit-learn/pull/18169

I have a couple of things that I did not get time to add before the merge to happen.

glemaitre commented:
The plot_det_curve was submitted there: https://github.com/scikit-learn/scikit-learn/pull/18176
